File: day1.py
#!/usr/bin/env python3

with open("day1.txt") as f:
    cals = [line.strip() for line in f]

# part 1
cur = 0
max_so_far = 0

for cal in cals:
    if cal == '':
        cur = 0
    else:
        cur += int(cal)
        max_so_far = max(max_so_far, cur)

print(max_so_far)

# part 2
cur = 0
elves = []

for cal in cals:
    if cal == '':
        elves.append(cur)
        cur = 0
    else:
        cur += int(cal)
elves.append(cur)
print(sum(sorted(elves)[::-1][:3]))

File: day1.ml
open! Core

let lines = In_channel.read_lines "day1.txt"

let by_elf =
  let elves, elf =
    List.fold lines ~init:([], 0) ~f:(fun (elves, cur_elf) line ->
        match line with
        | "" -> (cur_elf :: elves, 0)
        | calories ->
            let calories = Int.of_string calories in
            (elves, cur_elf + calories))
  in
  elf :: elves |> List.rev

let part1 () =
  List.max_elt by_elf ~compare:[%compare: int]
  |> Option.value_exn |> printf "%d"

let%expect_test _ =
  part1 ();
  [%expect {| 67027 |}]

let part2 () =
  let sorted = List.sort by_elf ~compare:[%compare: int] |> List.rev in
  match sorted with
  | a :: b :: c :: _ -> printf "%d %d %d" a b c
  | _ -> failwith "no"

let%expect_test _ =
  part2 ();
  [%expect {| 67027 65333 64931 |}]
